### PR TITLE
Fix documentation for {{each}} and {{with}} helpers

### PR DESCRIPTION
**Pull request by YoranBrondsema**
Addresses https://github.com/emberjs/ember.js/issues/10994.
## {{each}} helper

Restored the original documentation from 1.11 with the following changes:
- Updated to block params syntax (e.g. replace `{{#each post in posts}}` with `{{#each posts as |post|}}`
- Removed mentions of `ObjectController` and replace with `Controller`.
- Replaced `{{crossLink}}` helpers with links to relevant documentation.
## {{with}} helper
- Updated to block params syntax
- Expanded the example for the section about the `controller` option
## Comment about documentation of parameters

I still think in both cases the description of the parameters are unclear and I think this holds for all helpers. Like the `{{each}}` helper documents the parameters `name` and `path` but it's not clear how they're used and what they represent.

Instead, would it make sense to document the syntax of the helpers with a sort of grammar? I like how MDN documents the formal syntax of CSS rules. For instance, the formal syntax of the `box-shadow` rule is as follows (https://developer.mozilla.org/en-US/docs/Web/CSS/box-shadow#Formal_syntax):

```
none | [inset? && [ <offset-x> <offset-y> <blur-radius>? <spread-radius>? <color>? ] ]#
```

Using this for helpers you could have something as follows for `{{each}}`:

```
{{each <array> [itemViewClass="<item-view-class>"]? [emptyViewClass="<empty-view-class>"]? [itemController="<item-controller>"]? as |<item>|}}
```

In any case, correct me if I'm wrong but I don't think that in their current format, the documentation of the parameters are particularly useful to most of the Ember.js users.

### CODE REVIEW

File: packages/ember-htmlbars/lib/helpers/each.js [hunk 1]
```diff
@@ -2,6 +2,151 @@ import { get } from "ember-metal/property_get";
...
+  ```
+
+  ### Specifying an alternative view for each item
+
+  `itemViewClass` can control which view will be used during the render of each
+  item's template. The following template:
+
+  ```handlebars
+  <ul>
+    {{#each developers itemViewClass="person" as |developer|}}
```
Comment (mixonic) [inline packages/ember-htmlbars/lib/helpers/each.js]: We should probably not doc this, as it creates poor performance and will be removed.

File: packages/ember-htmlbars/lib/helpers/each.js [hunk 2]
```diff
@@ -2,6 +2,151 @@ import { get } from "ember-metal/property_get";
...
+  ```javascript
+  App.PersonView = Ember.View.extend({
+    tagName: 'li',
+    template: '{{developer.name}}'
+  });
+  ```
+
+  ```handlebars
+  <ul>
+    {{each developers itemViewClass="person" as |developer|}}
```
Comment (mixonic) [inline packages/ember-htmlbars/lib/helpers/each.js]: OMG, this makes my brain hurt so bad :-)

File: packages/ember-htmlbars/lib/helpers/each.js [hunk 3]
```diff
@@ -2,6 +2,151 @@ import { get } from "ember-metal/property_get";
...
+
+  ### Specifying an alternative view for no items (else)
+
+  The `emptyViewClass` option provides the same flexibility to the `{{else}}`
+  case of the each helper.
+
+  ```javascript
+  App.NoPeopleView = Ember.View.extend({
+    tagName: 'li',
+    template: 'No person is available, sorry'
```
Comment (mixonic) [inline packages/ember-htmlbars/lib/helpers/each.js]: Not a valid template

File: packages/ember-htmlbars/lib/helpers/each.js (hunk 1 above)
Comment (rwjblue) [inline packages/ember-htmlbars/lib/helpers/each.js]: It is also deprecated as of 1.13.

**Comment by mixonic** [comment]
@YoranBrondsema this is an improvement. JSCS is complaining about some whitespace in the code doc blocks (you can find the failure locally by running the tests).

Can we remove all mention of setting custom controllers and views? These APIs, in glimmer, are much slower than just using a component in side the each block, for instance. I think this is a good chance flag them in the documentation as legacy.

@rwjblue are the `itemViewClass` and `controller=` APIs going to be deprecated in the 1.13.x cycle?

**Comment by rwjblue** [comment]
@mixonic - Anything that opts you into the `LegacyEachView` was deprecated in https://github.com/emberjs/ember.js/pull/11129.

File: packages/ember-htmlbars/lib/helpers/each.js (hunk 1 above)
Comment (YoranBrondsema) [inline packages/ember-htmlbars/lib/helpers/each.js]: OK I removed this part of the documentation.

**Comment by YoranBrondsema** [comment]
OK I removed the documentation for `itemController`, `itemViewClass` and `emptyViewClass` for `{{each}}`.

Also fixed the trailing whitespace.

For `{{with}}`, now I'm not sure whether the `controller` parameter should stay or go?

**Comment by rwjblue** [comment]
`controller` should be removed from documentation (it will be deprecated in 1.13 just hasn't happened yet).

File: .gitignore [hunk 4]
```diff
@@ -39,3 +39,4 @@ publish_to_bower/
 bower_components/
 npm-debug.log
 .ember-cli
+*.swp
```
Comment (rwjblue) [inline .gitignore]: This should generally go in a global `.gitignore`. See [here](https://help.github.com/articles/ignoring-files/#create-a-global-gitignore) for info...

Comment (YoranBrondsema) [inline .gitignore]: That makes sense, thanks.

**Comment by mixonic** [comment]
restarted the build

**Comment by mixonic** [comment]
@YoranBrondsema thank you!!

**Pull request merged**